[v5] Add JSDoc-like comments to StudyUtils.jsm

biancadanforth opened the pull request:
Ready for review!

gregglind commented on src/StudyUtils.in.jsm:
```diff
@@ -22,6 +44,10 @@ const CID = Cu.import("resource://gre/modules/ClientID.jsm", null);
+/**
+* Gets the telemetry client ID for the user (AKA their Firefox user ID).
+* @return {string} id - the telemetry client ID
+*/
```
Also mark as `@async`

gregglind commented on src/StudyUtils.in.jsm:
```diff
@@ -1,5 +1,27 @@
+* OUTSTANDING QUESTIONS FOR GREGG
+*  - shield-studies-addon schema description references attributes sent as 
+*  Map(string, string). Can you explain what this means?
+*  - Can you describe the purpose of the `studySetup` and `webExtensionMsg` schema?
```
They are the validators for what a valid  `.setup` options, and a what is a valid message that the utils can handle from the bound method from a `webExtension` can send.

biancadanforth replied:
So just to confirm: `studySetup` schema validates the argument passed into `StudyUtils.setup()` and `webExtensionMsg` should be used to validate the first argument passed into `respondToWebExtensionMessage` in `StudyUtils.constructor`, but it doesn't look like it is used currently? Can you confirm?

gregglind commented on src/StudyUtils.in.jsm:
```diff
@@ -1,5 +1,27 @@
+/**
+* OUTSTANDING QUESTIONS FOR GREGG
+*  - shield-studies-addon schema description references attributes sent as 
+*  Map(string, string). Can you explain what this means?
```
an object with string keys and string values.

biancadanforth replied:
Ok thanks; do you explain anywhere what the attributes key is used for? I know it's for study-specific, string-string, key-value pairs -- looking at [Kamyar's CFR study](https://github.com/raymak/cfr-focused-release/blob/master/TESTPLAN.md), it looks like they are used to capture different, feature-related state information?

gregglind commented on src/StudyUtils.in.jsm:
```diff
@@ -22,6 +44,10 @@ const CID = Cu.import("resource://gre/modules/ClientID.jsm", null);
 const { TelemetryEnvironment } = Cu.import("resource://gre/modules/TelemetryEnvironment.jsm", null);
 
+/**
+* Gets the telemetry client ID for the user (AKA their Firefox user ID).
```
Don't call it a Firefox User ID.  WE do not track users, and they do not have IDs.  It is a Telemetry Client ID

mythmon replied:
Besides the optics issue, there are other uncorrelated IDs that clients have (like for Shield and Pioneer)

biancadanforth replied:
@mythmon , what do you mean by "optics"? I removed the aside. For my own understanding, what is the telemetry client ID based on? Is it unique to the user's browser?

gregglind commented on src/StudyUtils.in.jsm:
```diff
@@ -170,6 +285,7 @@ class StudyUtils {
     }.bind(this);
 
     // expose the sample utilities
+    // @QUESTION: What are each of these?
```
document thes on their functions.  
sha256: async, string into sha256 hash hexadecimal representation
cumsum => cumulative sum array of length n, to array of length n such that n_i = sum(array.slice(0,i)).  here used for turning weighted variations into right hand-limits for the sample choice bucket cutoffs.   [.25,.3,.45] =>  [.25,.55,1.0]
chooseWeighted:  weightedVariations, fraction=rand.random() =>  given a samplie weights and a particlar position, return the variation.  If no fraction given, return a variation at random proportionate to the weightVariations object.
hashFraction(saltedString, bit12) => return Number in [0,1).  convert a string into a Number in [0,1) based on the first <bits> bits  of it's sha256 hex represtation.

gregglind replied:
Glibly: The 'salt' ensures that a user ends up wth differnt bucket / hash / magic fraction for each study.
the HASHING makes sure that every bucket / variation gets filled.

gregglind commented on src/StudyUtils.in.jsm:
```diff
@@ -197,24 +323,48 @@ class StudyUtils {
+  /**
+  * @async
+  * TODO bdanforth: add docblock
+  * @QUESTIONS: Why is there a ShieldUtils method and a module function for `getTelemetryId`?
```
No good reason.

gregglind replied:
(maybe used by smapling stuff which is outside the class?)

biancadanforth replied:
I did a ctrl+F and I don't see it called anywhere except inside StudyUtils methods. Okay to move it in?

gregglind commented on src/StudyUtils.in.jsm:
```diff
@@ -197,24 +323,48 @@ class StudyUtils {
   }
+  /**
+  * TODO bdanforth: add docblock
+  * @QUESTIONS: Where is `setVariation` and `getVariation` called?
```
in bootstrap.js.

gregglind commented on src/StudyUtils.in.jsm:
```diff
@@ -225,6 +375,17 @@ class StudyUtils {
+  /**
+  * @async
+  * TODO bdanforth: add docblock
+  * @Question: what does this do? Ensure the same user gets the same branch every time?
```
Yes same variation each time.

gregglind commented on src/StudyUtils.in.jsm:
```diff
@@ -289,6 +461,10 @@ class StudyUtils {
+
+  /**
+  * @async
+  */
```
I hate this method, but it's a core feature of Study Utils.  See "about.md" and "survival"

gregglind commented on src/StudyUtils.in.jsm:
```diff
@@ -402,6 +588,9 @@ class StudyUtils {
   // telemetry from addon, mostly from webExtension message.
+  /**
+  * @async
+  */
```
telemetry (no undercores) does schema checking.

biancadanforth replied:
looks like `telemetry` forwards the data from the addon to `_telemetry` where the schema is validated

pdehaan commented:
At the risk of going offtopic and derailing everything... there are also two interesting ESLint rules for working w/ JSDoc, [**require-jsdoc**](https://eslint.org/docs/rules/require-jsdoc) and [**valid-jsdoc**](https://eslint.org/docs/rules/valid-jsdoc):

```
➜  shield-studies-addon-utils git:(jsdocify-shield-utils) $ npm run eslint

> shield-studies-addon-utils@4.1.0 eslint /Users/pdehaan/dev/github/mozilla/shield-studies-addon-utils
> eslint src --ext jsm --ext js --ext json

/Users/pdehaan/dev/github/mozilla/shield-studies-addon-utils/src/StudyUtils.in.jsm
  133:3  warning  Missing JSDoc comment                             require-jsdoc
  162:1  warning  Missing JSDoc parameter type for 'Object'         valid-jsdoc
  228:1  warning  Missing JSDoc @returns for function               valid-jsdoc
  361:3  warning  Missing JSDoc @returns for function               valid-jsdoc
  361:3  warning  Missing JSDoc for parameter 'url'                 valid-jsdoc
  361:3  warning  Missing JSDoc for parameter 'params'              valid-jsdoc
  394:3  warning  Missing JSDoc @returns for function               valid-jsdoc
  394:3  warning  Missing JSDoc for parameter 'variation'           valid-jsdoc
  408:3  warning  Missing JSDoc @returns for function               valid-jsdoc
  408:3  warning  Missing JSDoc for parameter 'weightedVariations'  valid-jsdoc
  408:3  warning  Missing JSDoc for parameter 'rng'                 valid-jsdoc
  479:3  warning  Missing JSDoc @returns for function               valid-jsdoc
  491:3  warning  Missing JSDoc @returns for function               valid-jsdoc
  541:3  warning  Missing JSDoc @returns for function               valid-jsdoc
  562:3  warning  Missing JSDoc @returns for function               valid-jsdoc
  562:3  warning  Missing JSDoc for parameter 'data'                valid-jsdoc
  562:3  warning  Missing JSDoc for parameter 'bucket'              valid-jsdoc
  618:3  warning  Missing JSDoc @returns for function               valid-jsdoc
  618:3  warning  Missing JSDoc for parameter 'data'                valid-jsdoc
  638:1  warning  Missing JSDoc comment                             require-jsdoc

✖ 20 problems (0 errors, 20 warnings)
```

```diff
diff --git a/.eslintrc.js b/.eslintrc.js
index eb773fc..6a9b91b 100644
--- a/.eslintrc.js
+++ b/.eslintrc.js
@@ -11,7 +11,7 @@ module.exports = {

   rules: {
@@ -27,5 +27,8 @@ module.exports = {
     "no-unused-vars": "error",
     "prefer-const": "warn",
     "semi": ["error", "always"],
+
+    "require-jsdoc": "warn",
+    "valid-jsdoc": "warn",
   },
 };
```

gregglind commented on src/StudyUtils.in.jsm:
```diff
@@ -1,5 +1,33 @@
+/**
+* STUDYUTILS OVERVIEW
+* This module takes care of the following:
+*  - Validates telemetry packets via JSON schema before sending pings
```
for these bucket types:   'shield-study', 'shield-study-addon', 'shield-study-error'.  see docs/telemetry.md

biancadanforth commented:
This PR is ready for a review!

* @gregglind , I left your original comments/TODOs in place. Please let me know if any can be removed/updated.
* I added the js-doc eslint rules per @pdehaan 's good recommendation. I also added a `max-len` rule of 80 characters. Currently the module has zero eslint warnings/errors.
* If there is interest, I can create a follow-up issue to actually make a JSDoc out of this file.

Looking forward to any and all feedback!

pdehaan commented on src/StudyUtils.in.jsm:
```diff
@@ -76,52 +143,107 @@ function merge(source) {
+* @example [.25,.3,.45] => [.25,.55,1.0]; if a user's sample weight were .25,
+* they would fall into the left-most bucket
+* @param {Number[]} arr - An array of sample weights [0, 1)
+* @returns {Number[]} - A cumulative sum array of sample weights [0, 1)
```
Should these be `[0, 1)`, or `[0, 1]` (or `(0, 1)`)?
Seems like an inconsistent bracket pairing.

(I think there were 5 other instances of `[0, 1)` in this PR, so it may be a search-replace thing.)

biancadanforth replied:
This is a shorthand way of saying `0 <= value < 1`; these values primarily come from [Math.random()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random).

I've replaced `[0,1)` with the longer form `0 <= value < 1` to avoid confusion. In one of the cases (the output of `cumsum`), the value can actually equal 1, so I've updated that too.

pdehaan commented on src/StudyUtils.in.jsm:
```diff
@@ -197,50 +387,109 @@ class StudyUtils {
+  * Opens a new tab that loads a page with the specified URL.
+  * @param {string} url - the url of a page
+  * @param {Object} params - optional, see
+  * http://mdn.beonex.com/en/XUL/Method/addTab.html
```
Seems like a scary URL. 
How about https://developer.mozilla.org/docs/Mozilla/Tech/XUL/Method/addTab

pdehaan commented on src/StudyUtils.in.jsm:
```diff
@@ -289,6 +587,21 @@ class StudyUtils {
+  * @param {string} reason - The reason the study is ending, see
+  * schema.studySetup.json
+  * @param {string} fullname -  optional, the full name of the study state,
+  * see schema.studySetup.json
```
I'm not sure these two `@param` entries are correct. It looks like `endStudy()` takes a single Object parameter w/ two keys, not 2 distinct, named parameters.

biancadanforth replied:
You're right; I've updated this for the destructured object argument.

pdehaan commented on src/StudyUtils.in.jsm:
```diff
@@ -76,52 +143,107 @@ function merge(source) {
+* Note: This is ultimately used to make a hash of the user's telemetry clientID
+* and the study name.
+* @param {string} message - The message to convert.
+* @returns {string} hashHex - a hexadecimal, 256-bit hash
```
For `@returns`, I dont think you need the variable name.
If that's right, we have a few instances of this:

```sh
$ git grep '@returns {' | grep -v '} - ' | grep -v '{void}'

src/StudyUtils.in.jsm:* @returns {string} hashHex - a hexadecimal, 256-bit hash
src/StudyUtils.in.jsm:  * @returns {string} id - the telemetry client ID
src/StudyUtils.in.jsm:  * @returns {Object} variation - the study variation for this user
src/StudyUtils.in.jsm:  * @returns {Object} queryArgs - the query arguments for the study
src/StudyUtils.in.jsm:* @returns {Object} L - the Logger instance, see gre/modules/Log.jsm
```

biancadanforth replied:
It is optional; I've removed it for more clarity.

pdehaan commented:
Are we writing the JSDocs to be good citizens, or are we publishing these API docs somewhere?

gregglind commented:
Good catches...

1.  Beonex... grrr!
2.  Writing for good citizen, jsdoc training.  EOL / counting inventory of
our shield stuff :). By Austin i would like to finish all docs and turn off
active dev.  I am open to “publish”, if you have suggestions.



On Sat, Dec 2, 2017 at 1:33 AM Peter deHaan <notifications@github.com>
wrote:

> Are we writing the JSDocs to be good citizens, or are we publishing these
> API docs somewhere?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mozilla/shield-studies-addon-utils/pull/107#issuecomment-348674901>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKAj2cBXBFMagZPnVKesg2nJkA-KEgdks5s8P1fgaJpZM4QhOEo>
> .
>

gregglind commented:
I finally realize there is a word for Reals in [0, 1].   It’s a
probability.  Still no good word for Reals in [0, 1).

On Sat, Dec 2, 2017 at 10:59 AM Bianca Danforth <notifications@github.com>
wrote:

> *@biancadanforth* commented on this pull request.
> ------------------------------
>
> In src/StudyUtils.in.jsm
> <https://github.com/mozilla/shield-studies-addon-utils/pull/107#discussion_r154500839>
> :
>
> >    return hashHex;
>  }
>
> +/**
> +* Converts an array of length N into a cumulative sum array of length N,
> +* where n_i = sum(array.slice(0,i)) i.e. each element is the sum of all
> +* elements up to and including that element
> +* This is ultimately used for turning sample weights (AKA weightedVariations)
> +* into right hand limits (>= X) to  deterministically select which variation
> +* a user receives.
> +* @example [.25,.3,.45] => [.25,.55,1.0]; if a user's sample weight were .25,
> +* they would fall into the left-most bucket
> +* @param {Number[]} arr - An array of sample weights [0, 1)
> +* @returns {Number[]} - A cumulative sum array of sample weights [0, 1)
>
> This is a shorthand way of saying 0 <= value < 1; these values primarily
> come from Math.random()
> <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random>
> .
>
> I've replaced [0,1) with the longer form 0 <= value < 1 to avoid
> confusion. In one of the cases (the output of cumsum), the value can
> actually equal 1, so I've updated that too.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/mozilla/shield-studies-addon-utils/pull/107#discussion_r154500839>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKAj3o9dJclzDfLPM1RK4B22KYW3QZWks5s8YHxgaJpZM4QhOEo>
> .
>

biancadanforth commented:
Regarding publishing, I've no direct experience, but @rhelmer recommends [sphinx-js](https://hacks.mozilla.org/2017/07/introducing-sphinx-js-a-better-way-to-document-large-javascript-projects/). He just landed sphinx-js support in http://firefox-source-docs.mozilla.org/, and he says readthedocs likely has some opt-in beta-level support for sphinx-js. This would be a follow-up issue I'm happy to work on next if desired, though I would not get to it by the AllHands or even the end of this year in all likelihood.

Barring any other feedback; this PR, however, is ready to go! :)

pdehaan commented:
I've never published JSDoc content personally, but I know the people on **mozilla-lockbox** are doing it in a couple repos; see [**mozilla-lockbox/lockbox-extension** /package.json:28-29](https://github.com/mozilla-lockbox/lockbox-extension/blob/0f433b4793c596647bdf6f1d8aafac11eb3552b0/package.json#L28-L29):

```json
    "doc": "documentation build src/webextension/** -f md -o docs/api.md",
    "postdoc": "mkdocs build",
```

Looks like the [**documentation**](http://npm.im/documentation) is a Node based thing, and the other half [**mkdocs**](http://www.mkdocs.org/) seems to be Python, which they're invoking from a **postdoc** npm hook above, and from Travis-CI.
If I get bored next week, I'll try poking at **documentation** module to see if I can get some API docs built locally and see how they render.

rhelmer commented:
@erikrose (the author of https://github.com/erikrose/sphinx-js) describes the readthedocs situation here: https://github.com/erikrose/sphinx-js#example

I'd second the suggestion in the document above to use https://github.com/fathom as an example of how to use sphinx-js. You can also read more about how Firefox uses sphinx https://mail.mozilla.org/pipermail/firefox-dev/2017-November/005955.html

Hope that helps!

gregglind commented:
(Wombat numbers.
https://math.stackexchange.com/questions/2489/is-there-a-name-for-0-1 )

On Sat, Dec 2, 2017 at 3:47 PM Robert Helmer <notifications@github.com>
wrote:

> @erikrose <https://github.com/erikrose> (the author of
> https://github.com/erikrose/sphinx-js) describes the readthedocs
> situation here: https://github.com/erikrose/sphinx-js#example
>
> I'd second the suggestion in the document above to use
> https://github.com/fathom as an example of how to use sphinx-js. You can
> also read more about how Firefox uses sphinx
> https://mail.mozilla.org/pipermail/firefox-dev/2017-November/005955.html
>
> Hope that helps!
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/mozilla/shield-studies-addon-utils/pull/107#issuecomment-348722534>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKAj2jjB62P1oXq39qvHnSwRhKNd4tzks5s8cV1gaJpZM4QhOEo>
> .
>

motin commented:
This PR is related to https://github.com/mozilla/shield-studies-addon-utils/issues/105 (adding here to make it easier to find related issues)

motin commented:
Feel free to [change the base branch of this PR](https://blog.github.com/2016-08-15-change-the-base-branch-of-a-pull-request/) to `develop` and merge it, it will be the basis of v5 of utils for sure :)
